# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](https://www.datacamp.com/community/tutorials/tutorial-jupyter-notebook##UseJupyter).

Additionally, in order to write text in this document, you will be making use of Markdown cells. A handy cheatsheet for Markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1

Note: the exercise numbers below match up with the reading for ease of cross referencing.

### Exercise 1.4

Start the Python interpreter and use it as a simple calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *

* How many seconds are there in 42 minutes 42 seconds?
* How many miles are there in 10 kilometers? Hint: there are 1.61 kilometers in a mile.
* If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average pace (time per mile in minutes and seconds)? What is your average speed in miles per hour?

Please use the following markdown cell to provide your answers.

In [3]:
42*60+40

2560

In [5]:
10/1.61

6.211180124223602

In [6]:
2560/6.211

412.1719529866366

In [7]:
412/60

6.866666666666666

In [8]:
0.866666666666666*60

51.999999999999964

In [4]:
(10/1.61)/((42+42/60)/60)

8.727653570337614

Bullet point 1
---
seconds = minutes\*60 + seconds\_given

-> seconds = 42\*60 + 42

-> seconds = 2560

Bullet point 2
---
10 km \* 1 mile / 1.61 km = 6.21 miles

Bullet point 3 Part 1
---

2560 seconds / 6.21 miles = 412 seconds/mile

-> 6.866666 minutes 

-> 6 minutes + 0.866666 minutes

-> 6 minutes + 0.866666 minutes * 60 seconds / 1 minute

-> 6 minutes, 52 secons

Bullet point 3 Part 2
---
6.21 miles / (2560 seconds / (60 seconds/minute) / (60 minutes/hour))

-> 8.72 miles/hour

## Chapter 2

### Exercise

Assume that we execute the following assignment statements:

```
width = 17
height = 12.0
delimiter = '.'
```

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

Example: `width * 2` <br/>
Answer: 34, integer

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

In [16]:
width = 17
height = 12.0
delimiter = '.'

number1 = width/2
print('1.')
print(number1)
print('Type:')
print(type(number1))
print('')

number2 = width/2.0
print('2.')
print(number2)
print('Type:')
print(type(number2))
print('')

number3 = height/3
print('3.')
print(number3)
print('Type:')
print(type(number3))
print('')

number4 = 1+2*5
print('4.')
print(number4)
print('Type:')
print(type(number4))
print('')

number5 = delimiter * 5
print('5.')
print(number5)
print('Type:')
print(type(number5))

1.
8.5
Type:
<class 'float'>

2.
8.5
Type:
<class 'float'>

3.
4.0
Type:
<class 'float'>

4.
11
Type:
<class 'int'>

5.
.....
Type:
<class 'str'>


### Exercise 2.2

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

The volume of a sphere with radius r is $4/3 \pi  r^3$. What is the volume of a sphere with radius 5? <br />
Hint: 392.7 is wrong! You can approximate pi with 3.14 or import it through the Python math library.

In [18]:
pi = 3.14 #pi
r = 5 #radius
Volume = pi*4/3 * pi * r**3
print(Volume)

1643.2666666666667


Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [20]:
cover_price = 24.95
after_discount = 0.6 #40% discount leaves 60% paid
shipping_initial = 3
shipping_each = 0.75
copies = 60

book_cost = cover_price*after_discount*copies
shipping_cost = shipping_initial + shipping_each*(copies-1)

wholesale_cost = book_cost + shipping_cost
print(wholesale_cost)

945.4499999999999


If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [23]:
easy_miles = 2
tempo_miles = 3
easy_pace_min = 8
easy_pace_sec = 15
tempo_pace_min = 7
tempo_pace_sec = 12
min_sec = 60
hour_min = 60


easy_pace_total = easy_pace_min + easy_pace_sec/min_sec #total easy pace in miles/minute
tempo_pace_total = tempo_pace_min + tempo_pace_sec/min_sec 

easy_time = easy_miles*easy_pace_total
tempo_time = tempo_miles*tempo_pace_total
total_time_minutes = easy_time+tempo_time

time_start = 6*hour_min+52
time_end = time_start + total_time_minutes
minutes_end = 52 + total_time_minutes
#hours_end = floor(time_end/hour_min)
#minutes_end = time_end - hours_end*hour_min

#print(hours_end*100+minutes_end)
print(time_end)
print(minutes_end)

450.1
90.1


the second number there is how many minutes past 6 AM you get back home.  Therefore, you get home at 7:30 AM

## Chapter 3

### Exercise 3.1

Write a function named `right_justify` that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('monty')
                                                                 monty
```
Hint: Use string concatenation and repetition. Also, Python provides a built-in function called `len` that returns the length of a string, so the value of `len('monty')` is 5.

In [27]:
import math
def right_justify(s):
    single_space = ' '
    num_spaces = 70-len(s)
    print(single_space*num_spaces + s)
    
right_justify('monty')


                                                                 monty


### Exercise 3.3

Note: This exercise should be done using only the statements and other features we have learned so far.

1\. Write a function that draws a grid like the following:

```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
Note: There are four empty spaces between the vertical bars.

To print more than one value on a line, you can print a comma-separated sequence of values:

```
print('+', '-')
```

By default, `print` advances to the next line, but you can override that behavior and put a space at the end, like this:

```
print('+', end=' ')
print('-')
```

The output of these statements is '+ -'.

A `print` statement with no argument ends the current line and goes to the next line.

In [57]:
def do_four(x):
    x()
    x()
    x()
    x()
   
    
bar = "- - - - + "
column = "        | "
num_grids = 2 #number of grids

def make_column():
    print("| " + column*2)
    
def make_bar():
    print("+ " + bar*2)

def make_basket():
    do_four(make_column)
    make_bar()

    
make_bar()
make_basket()
make_basket()



+ - - - - + - - - - + 
|         |         | 
|         |         | 
|         |         | 
|         |         | 
+ - - - - + - - - - + 
|         |         | 
|         |         | 
|         |         | 
|         |         | 
+ - - - - + - - - - + 


2\. Write a function that draws a similar grid with four rows and four columns.

In [59]:
def do_four(x):
    x()
    x()
    x()
    x()
   
    
bar = "- - - - + "
column = "        | "
num_grids = 4 #number of grids

def make_column():
    print("| " + column*num_grids)
    
def make_bar():
    print("+ " + bar*num_grids)

def make_basket():
    do_four(make_column)
    make_bar()

    
make_bar()
do_four(make_basket)

+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 


## Chapter 5.1 - 5.7

### Exercise 5.2

Fermat’s Last Theorem says that there are no positive integers $a$, $b$, and $c$ such that $$a^n + b^n = c^n$$ for any values of $n$ greater than 2.

1\. Write a function named `check_fermat` that takes four parameters – $a$, $b$, $c$ and $n$ – and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$, then the program should print: "Holy smokes, Fermat was wrong!" Otherwise the program should print: "No, that doesn’t work."

In [62]:
def check_fermat(a,b,c,n):
    if a**n + b**n == c**n and n>2:
        print("Holy smokes, Fermat was wrong!")
    else: 
        print("No, that doesn't work.")
    
check_fermat(3,4,5,2)

No, that doesn't work.


2\. Write a function that prompts the user to input values for $a$, $b$, $c$ and $n$, converts them to integers, and uses `check_fermat` to check whether they violate Fermat’s theorem.

In [65]:
def check_fermat(a,b,c,n):
    if a**n + b**n == c**n and n>2:
        print("Holy smokes, Fermat was wrong!")
    else: 
        print("No, that doesn't work.")
        
def ask_fermat():
    prompta = "What is a?\n"
    promptb = "what is b?\n"
    promptc = "what is c?\n"
    promptn = "what is n?\n"
    
    a = input(prompta)
    b = input(promptb)
    c = input(promptc)
    n = input(promptn)
    
    check_fermat(int(a),int(b),int(c),int(n))

ask_fermat()

What is a?
1
what is b?
2
what is c?
3
what is n?
4
No, that doesn't work.


### Exercise 5.3 

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

1\. Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [67]:
def is_triangle(a,b,c):
    if a+b<=c or a+c<=b or b+c<=a:
        print("No")
    else:
        print("Yes")

is_triangle(12,1,12)

Yes


2\. Write a function that prompts the user to input three stick lengths, converts them to integers, and uses `is_triangle` to check whether sticks with the given lengths can form a triangle.

In [69]:
def is_triangle(a,b,c):
    if a+b<=c or a+c<=b or b+c<=a:
        print("No")
    else:
        print("Yes")
        
def is_triangle_prompt():
    prompta = "What is a?\n"
    promptb = "what is b?\n"
    promptc = "what is c?\n"
    promptn = "what is n?\n"
    
    a = int(input(prompta))
    b = int(input(promptb))
    c = int(input(promptc))
    
    is_triangle(a,b,c)
    
is_triangle_prompt()
 

What is a?
1
what is b?
1
what is c?
3
No


## Quick poll
About how long did you spend working on this Reading Journal?

2 hours

## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.

I'm struggling with syntax but I think that's it.